### DISCUSSION TITLE
How to export trained agent in julia ReinforcementLearning

### DISCUSSION THREAD

**Discussion by Giotsik** [JuliaReinforcementLearning/ReinforcementLearning.jl#235] [Q&A]
Hello, I would like to know how to export a trained agent in this package, in order to run it for several experiment scenarios

**Comment by findmyway** [comment]
In case you haven't seen this:

https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/discussions/183#discussioncomment-286692

Because most models are constructed with Flux.jl, you can refer [Saving and Loading Models](https://fluxml.ai/Flux.jl/stable/saving/) in Flux.jl for more details.

**Reply by Giotsik** [reply]
I see, thanks for your reply! Could I have a structured example, if you have the ease, of the code on an application of a specific experiment where the agent used is an already trained one and is exported. I am actually working on my thesis and spend several time training a DQN agent for a set of motors, thus I would like to load it in the real machine system as a controller. Thanks for the valuable help. Regards

**Reply by findmyway** [reply]
Here's the example of the most flexible version with BSON.jl that is compatible across different versions of RL.jl:

```julia
using ReinforcementLearning
ex = E`JuliaRL_BasicDQN_CartPole`
policy = ex.policy;

using Flux: params
ps = params(policy);  # if your policy is on GPU, you need to transfer it to CPU first by calling `gpu`

using BSON
BSON.@save "ps.bson" ps

# Now you have saved parameters on your disk
```

```julia
# Next we load it in a new experiment
using ReinforcementLearning
ex = E`JuliaRL_BasicDQN_CartPole`
policy = ex.policy;

using BSON
BSON.@load "ps.bson" ps

using Flux
Flux.loadparams!(policy, ps)  # remember to move ps to GPU if your policy runs on GPU
```

Maybe I should add this into the [Guide](https://juliareinforcementlearning.org/guide/)

**Reply by Giotsik** [reply]
Thank you that seems to work quite perfect!

**Reply by findmyway** [reply]
Cool. Glad it helps.

**Reply by Giotsik** [reply]
![image](https://user-images.githubusercontent.com/75027633/114882295-7bdfe900-9e0c-11eb-974f-22e4ad2e4013.png)
Hey sorry one more thing, in Pluto it seems to have this error, is it because of the gpu-cpu thing you mentioned? plus when I save it with cpu() it seems to show an error of not identifying the model 
![image](https://user-images.githubusercontent.com/75027633/114882650-d5481800-9e0c-11eb-9c76-8e478ff8f0f6.png)
regards

**Reply by findmyway** [reply]
For the first question, you need to install `Zygote` first and `using Zygote` before loading parameters.
For the second question, it seems your code is invalid here.

Try `ps = cpu(params(gdpolicy))` instead?

**Reply by Giotsik** [reply]
yeah sorry about the second. About the first I had already the package added and been using it :(
![image](https://user-images.githubusercontent.com/75027633/114926202-beb7b600-9e38-11eb-9b8c-52c6795e1b48.png)

**Reply by findmyway** [reply]
That's strange. Could you check if your BSON is the latest version and try again in REPL instead of Pluto?
